src/librustc_mir/interpret/operand.rs

```diff
@@ -684,12 +707,16 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                                 .variants
                                 .len();
                             assert!(usize::try_from(variant_index).unwrap() < variants_len);
-                            (u128::from(variant_index), VariantIdx::from_u32(variant_index))
+                            VariantIdx::from_u32(variant_index)
                         } else {
-                            (u128::from(dataful_variant.as_u32()), dataful_variant)
+                            dataful_variant
                         }
                     }
-                }
+                };
+                // Compute the size of the scalar we need to return.
+                // FIXME: Why do we not need to do a cast here like we do above?
```

Comment:
This one particularly confused me -- we should resolve that question before landing.

Reply:
For the niche layout, the variant indices have to not overflow host `u32` *nor* the target (type-checking) discriminant type (`isize` by default but I guess it could be `u8`/`i8`).

So you can add an assert that it fits as a positive number in the discriminant type, but no cast should be necessary.

Reply:
`Scalar::from_u32` already has such an assert (at least on debug builds).